File: src/main/java/com/lessmarkup/userinterface/model/structure/TypeaheadModel.java
package com.lessmarkup.userinterface.model.structure;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import com.google.inject.Inject;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.recordmodel.PropertyCollectionManager;
import com.lessmarkup.interfaces.structure.CachedNodeInformation;
import com.lessmarkup.interfaces.structure.ChildHandlerSettings;
import com.lessmarkup.interfaces.structure.NodeCache;
import com.lessmarkup.interfaces.structure.NodeHandler;
import com.lessmarkup.interfaces.structure.Tuple;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

public class TypeaheadModel {
    
    private final DataCache dataCache;
    private final DomainModelProvider domainModelProvider;
    private List<String> records;

    @Inject
    public TypeaheadModel(DataCache dataCache, DomainModelProvider domainModelProvider) {
        this.dataCache = dataCache;
        this.domainModelProvider = domainModelProvider;
    }
    
    
    public void initialize(String path, String property, String searchText) {
        NodeCache nodeCache = this.dataCache.get(NodeCache.class);
        
        Tuple<CachedNodeInformation,String> node = nodeCache.getNode(path);
        
        if (node == null) {
            throw new IllegalArgumentException("Cannot find node");
        }
        
        NodeHandler handler = DependencyResolver.resolve(node.getValue1().getHandlerType());
        
        List<PropertyCollectionManager> collectionManagers = new ArrayList<>();
        if (PropertyCollectionManager.class.isAssignableFrom(node.getValue1().getHandlerType())) {
            collectionManagers.add((PropertyCollectionManager) handler);
        }
        
        String rest = node.getValue2();
        
        while (rest != null && rest.length() > 0) {
            ChildHandlerSettings childSettings = handler.getChildHandler(rest);
            if (childSettings == null || !childSettings.getId().isPresent()) {
                throw new IllegalArgumentException();
            }
            handler = childSettings.getHandler();
            if (PropertyCollectionManager.class.isAssignableFrom(handler.getClass())) {
                collectionManagers.add((PropertyCollectionManager) handler);
            }
            
            rest = childSettings.getRest();
        }
        
        if (!collectionManagers.isEmpty()) {
            try (DomainModel domainModel = this.domainModelProvider.create()) {
                for (PropertyCollectionManager manager : collectionManagers) {
                    List<String> collection = manager.getCollection(domainModel, property, searchText);
                    if (collection == null || collection.isEmpty()) {
                        continue;
                    }
                    
                    records = new ArrayList<>();
                    collection.stream().limit(10).forEach(records::add);
                    return;
                }
            } catch (Exception ex) {
                Logger.getLogger(TypeaheadModel.class.getName()).log(Level.SEVERE, null, ex);
            }
        }
    }
    
    public JsonObject toJson() {
        JsonObject builder = new JsonObject();
        JsonArray array = new JsonArray();
        if (this.records != null) {
            for (String record : records) {
                array.add(new JsonPrimitive(record));
            }
        }
        builder.add("records", array);
        return builder;
    }
}

File: src/main/scala/com/lessmarkup/userinterface/model/structure/TypeaheadModel.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.model.structure

import com.google.gson.JsonObject
import com.google.inject.Inject
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.data.DomainModelProvider
import com.lessmarkup.interfaces.structure.NodeCache

class TypeaheadModel @Inject() (dataCache: DataCache, domainModelProvider: DomainModelProvider) {
  //private var records: List[String] = null

  def initialize(path: String, property: String, searchText: String) {
    val nodeCache: NodeCache = this.dataCache.get(classOf[NodeCache])
    val node = nodeCache.getNode(path)
    if (node.isEmpty) {
      throw new IllegalArgumentException("Cannot find node")
    }
    /*var handler: NodeHandler = DependencyResolver.resolve(node._1.getHandlerType)
    val collectionManagers: List[PropertyCollectionManager] = new ArrayList[PropertyCollectionManager]
    if (classOf[PropertyCollectionManager].isAssignableFrom(node._1.getHandlerType)) {
      collectionManagers.add(handler.asInstanceOf[PropertyCollectionManager])
    }
    var rest: String = node._2
    while (rest != null && rest.length > 0) {
      val childSettings: ChildHandlerSettings = handler.getChildHandler(rest)
      if (childSettings == null || !childSettings.getId.isPresent) {
        throw new IllegalArgumentException
      }
      handler = childSettings.getHandler
      if (classOf[PropertyCollectionManager].isAssignableFrom(handler.getClass)) {
        collectionManagers.add(handler.asInstanceOf[PropertyCollectionManager])
      }
      rest = childSettings.getRest
    }
    if (!collectionManagers.isEmpty) {
      try {
        val domainModel: DomainModel = this.domainModelProvider.create
        try {
          import scala.collection.JavaConversions._
          for (manager <- collectionManagers) {
            val collection: List[String] = manager.getCollection(domainModel, property, searchText)
            if (collection == null || collection.isEmpty) {
              continue //todo: continue is not supported
            }
            records = new ArrayList[String]
            collection.stream.limit(10).forEach(records.add)
            return
          }
        }
        catch {
          case ex: Exception => {
            Logger.getLogger(classOf[TypeaheadModel].getName).log(Level.SEVERE, null, ex)
          }
        } finally {
          if (domainModel != null) domainModel.close()
        }
      }
    }*/
  }

  def toJson: JsonObject = {
    val builder: JsonObject = new JsonObject
    /*val array: JsonArray = new JsonArray
    if (this.records != null) {
      import scala.collection.JavaConversions._
      for (record <- records) {
        array.add(new JsonPrimitive(record))
      }
    }
    builder.add("records", array)*/
    builder
  }
}